Refactor Node/Store Admin Pages

mrtracy opened the pull request:
Initial set of refactors for node/store status pages in the admin UI:
- Remove unused "historical status" functionality.
- Commonizes most of the data manipulation functions used by Store and Node
  models
- Moves Store and Node models to a shared Status file

There is still refactoring to be done after this change; for example, there is
still view code in the model.  However, this change is already becoming
considerably complex.

This commit also fixes a bug in the Makefile where the Typescript was never
rebuilt.

Commit:
Refactor Node/Store Admin Pages Pt. I

Initial set of refactors for node/store status pages in the admin UI:
+ Remove unused "historical status" functionality.
+ Commonizes most of the data manipulation functions used by Store and Node
models
+ Moves Store and Node models to a shared Status file

There is still refactoring to be done after this change; for example, there is
still view code in the model.  However, this change is already becoming
considerably complex.

This commit also fixes a bug in the Makefile where the Typescript was never
rebuilt.

tschottdorf commented on resource/ui/ts/models/status.ts:
```diff
@@ -0,0 +1,218 @@
+            return Math.floor(status.replicated_range_count / status.leader_range_count * 100).toString() + "%";
+        }
+
+        var _datetimeFormater = d3.time.format("%Y-%m-%d %H:%M:%S");
```
s/Formater/Formatter/

BramGruneir commented on resource/Makefile:
```diff
@@ -41,7 +41,7 @@ all: $(GOBINDATA_TARGET)
 debug: $(GOBINDATA_DEBUG_TARGET)
 	rm -f $(GOBINDATA_TARGET)
 
-$(TYPESCRIPT_TARGET): $(TYPESCRIPT_SOURCES)
```
What does this change do?

mrtracy replied:
As it was, TYPESCRIPT_TARGET (app.js) was only being rebuilt if the actual directory ui/ts changed somehow.  This change makes it so that app.js is rebuilt if any file under the ts directory changes.

tamird replied:
the old version didn't depend on the files in the directory, so it would not recompile them when they changed.

BramGruneir replied:
Cool.

BramGruneir commented on resource/ui/ts/models/proto.ts:
```diff
@@ -37,6 +37,48 @@ module Models {
+         * Create a new object which implements MVCCStats interface, with zero
+         * values.
+         */
+        export function NewMVCCStats():MVCCStats {
```
Perhaps we should move the proto interfaces to a different file from the implementations.  
Put the interfaces into /typeings/proto.ts.
And any implementation into their own model files.  If everything is in a single large file, than it will get too busy.

mrtracy replied:
Maybe.  I've got a lot of forward work on the next part of this refactoring, i'm trying to break it up into understandable chunks; i'll add a ToDo in my current PR, if that's alright.

BramGruneir replied:
A ToDo works for me.

BramGruneir commented on resource/ui/ts/models/proto.ts:
```diff
@@ -113,6 +155,32 @@ module Models {
+        /**
+         * Status is the common interface shared by NodeStatus and StoreStatus.
+         */
+        export interface Status{
```
Should node status and store status extend Status then?

mrtracy replied:
They can but it's not explicitly necessary, it will be asserted any time you try to assign a NodeStatus to something that expects a Status.

mrtracy replied:
I've addressed this in my next PR.

mrtracy replied:
Well actually, "Status" isn't going to be generated if we move to a generation model.   Along with the various static methods I've added, "Status" should probably not be in that file (as you mentioned on your earlier comment, perhaps the generated protos should go into the typings folder, or at least a separate file).

So these types should not extend from Status.

BramGruneir replied:
Are we moving to a generated model?  Do we need to write the generator ourselves or did you find one that works? My searches were fruitless.

It just seems very strange to have such a backwards relationship.  I have no problem with status, but we should have a function that creates a status from a node status and a store status.

mrtracy replied:
There is no need, you can just assign a NodeStatus to a Status variable and it works.  It's duck typed.

BramGruneir replied:
That will always unnerve me a bit.

mrtracy replied:
Do the Interfaces in Go unnerve you? :)

BramGruneir replied:
A little bit. I prefer explicit interfaces.  But at least with TS and Go, we get static analysis to tell us when we're wrong.

BramGruneir commented on resource/ui/ts/models/status.ts:
```diff
@@ -0,0 +1,218 @@
+                        });
+                        return results;
+                    });
+            }
```
This pattern here is everywhere.  Any chance we can abstract it?

mrtracy replied:
I have done this as part of my next PR.

BramGruneir commented on resource/ui/ts/models/status.ts:
```diff
@@ -0,0 +1,218 @@
+
+module Models {
+    export module Status {
+        import promise = _mithril.MithrilPromise;
```
Should you have a sub module here for node and store instead of having them in the same module?

mrtracy replied:
I think this is plenty of namespacing.

BramGruneir replied:
ok

BramGruneir commented on resource/ui/ts/models/status.ts:
```diff
@@ -0,0 +1,218 @@
+            }
+        }
+
+        export interface NodeStatusResponseSet {
```
Everything being returned as an array is wrapped in a 'd'.  I think a generic interface for WrappedResponseSet<T> might make more sense than repeating this same thing so many times.

mrtracy replied:
This interface is dead simple, I don't think there's any reason not to simply repeat it where needed.  On that note, it probably shouldn't be exported from this package.

BramGruneir commented on resource/ui/ts/models/status.ts:
```diff
@@ -0,0 +1,218 @@
+         * nonJsonErrors ensures that error messages returned from the server
+         * are parseable as JSON strings.
+         */
+        function nonJsonErrors(xhr: XMLHttpRequest, opts: _mithril.MithrilXHROptions): string {
```
This is everywhere, add this to the query helper once you extract it.

mrtracy replied:
I'll extract a Query package with "Get" and "Put" methods.

Commit:
Admin Node/Status Page Refactor Pt. II

Commit pulls out a common "QueryCache<T>" which can be shared between time
series queries and node status queries.

Added 'title' property to Query, which is used by graphs.

Significant changes to Status and Node pages to take advantage of newly
refactored APIs.

mrtracy commented:
Went ahead and uploaded another separate commit, it should address some of the concerns you had in the previous PR.

BramGruneir commented:
Nice.  I'll take another pass at it.

BramGruneir commented on resource/ui/ts/pages/stores.ts:
```diff
@@ -14,166 +14,143 @@ module AdminViews {
-      }
-    }
+        /**
+         * NodesPage show a list of all the available nodes.
```
should be StoresPage

BramGruneir commented on resource/ui/ts/pages/stores.ts:
```diff
@@ -14,166 +14,143 @@ module AdminViews {
-        ]);
-      }
+        /**
+         * NodePage show the details of a single node.
```
StorePage

BramGruneir commented on resource/ui/ts/models/status.ts:
```diff
@@ -0,0 +1,227 @@
+// source: models/store_status.ts
```
models/status.ts

BramGruneir commented on resource/ui/ts/util/querycache.ts:
```diff
@@ -0,0 +1,130 @@
+         * Check for a completed outstanding query, replacing cached results
+         * with the result.
+         */
+        private processOutstanding() {
```
Why do you call this with every function instead of chaining it to when the query completes?

mrtracy replied:
I actually like mithril's promise model.  It depends on a redraw being triggered whenever a network call completes, but it means that very little code needs to be wrapped up in promises.

In this model, the only thing that promises should do is transform the result.

BramGruneir replied:
Ok, works for me.

BramGruneir commented:
Some comment nits and a few questions.  But overall, I really like how this is looking.

BramGruneir commented:
LGTM
